File: MapReduce/task1.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkConf
import org.json4s._
import org.json4s.jackson.JsonMethods._
import scala.collection.mutable._
import org.json4s.jackson.Serialization
import java.io._

object task1 {
  def main(args: Array[String]): Unit = {
    val review_filepath = args(0)
    val output_filepath = args(1)

    val conf = new SparkConf
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")
    val review = sc.textFile(review_filepath).map(line => parse(line)).cache()
    val result = HashMap[String,Any]()

    // A.
    result("n_review") = review.count()
    println(s"total number of reviews: ${result("n_review")}")

    // B.
    result("n_review_2018") = review.filter(r => compact(render(r \ "date"))regionMatches(false, 1, "2018", 0, 4)).count()
    println(s"total number of reviews in 2018: ${result("n_review_2018")}")

    // C.
    result("n_user") = review.map(r => compact(render(r \ "user_id"))).distinct().count()
    println(s"number of distinct users who have written the reviews: ${result("n_user")}")

    // D.
    def mapfunc(x: JValue, id: String): (String, Int) = {
      val str = compact(render(x \ id))
      (str.substring(1, str.length()-1), 1)
    }
    val topusers = review.map(x => mapfunc(x, "user_id")).reduceByKey(_+_).takeOrdered(10)(Ordering[(Int, String)].on(s => (-s._2, s._1)))
    val res:Array[List[Any]] = new Array(topusers.length)
    for (i <- 0 until topusers.length) {
      res(i) = topusers(i).productIterator.toList
    }
    result("top10_user") = res

    // E.
    result("n_business") = review.map(r => compact(render(r \ "business_id"))).distinct().count()

    //F.
    val topbusiness = review.map(x => mapfunc(x, "business_id")).reduceByKey(_+_).takeOrdered(10)(Ordering[(Int, String)].on(s => (-s._2, s._1)))
    val res1:Array[List[Any]] = new Array(topbusiness.length)
    for (i <- 0 until topbusiness.length) {
      res1(i) = topbusiness(i).productIterator.toList
    }
    result("top10_business") = res1

    println(result)

    implicit val formats = DefaultFormats
    val writer = new PrintWriter(new File(output_filepath))
    writer.write(Serialization.writePretty(result))
    writer.close()
  }
}

File: MapReduce/task1.py
import sys
from operator import add

from pyspark import SparkContext
import json
from datetime import datetime


# A. The total number of reviews
def get_total_num_review(review_ids_rdd):
    return review_ids_rdd.count()


#B. The number of reviews in certain year
def get_num_review_in_certain_year(review_date_rdd, year):
    return review_date_rdd.filter(lambda kv: datetime.strptime(
        kv[1], '%Y-%m-%d %H:%M:%S').year == year).count()


# C. The number of distinct users who have wrote reviews
def get_distinct_user(user_ids_rdd):
    return user_ids_rdd.distinct().count()


# D. Top m users who wrote the largest numbers of reviews and the number of reviews they wrote
def get_top_u_users(business_user_rdd, top_u):
    return business_user_rdd.map(lambda kv: (kv[1], 1)).reduceByKey(add). \
        takeOrdered(top_u, key=lambda kv: (-kv[1], kv[0]))


# E. The number of distinct businesses that have been reviewed
def get_distinct_businesses(businesses_ids_rdd):
    return businesses_ids_rdd.distinct().count()

# F. The top 10 businesses that had the largest numbers of reviews and the number of reviews they had
def get_top_b_businesses(user_business_rdd, top_b):
    return user_business_rdd.map(lambda kv: (kv[1], 1)).reduceByKey(add). \
        takeOrdered(top_b, key=lambda kv: (-kv[1], kv[0]))



if __name__ == '__main__':
    review_filepath = sys.argv[1]
    output_filepath = sys.argv[2]
    year = 2018
    top_u = 10
    top_b = 10
    
    sc = SparkContext.getOrCreate()
    result_dict = dict()

    reviews = sc.textFile(review_filepath).map(lambda row: json.loads(row))

    review_ids_rdd = reviews.map(lambda kv: kv['review_id'])
    result_dict['n_review'] = get_total_num_review(review_ids_rdd)

    review_date_rdd = reviews.map(lambda kv: (kv['review_id'], kv['date']))
    result_dict['n_review_2018'] = get_num_review_in_certain_year(review_date_rdd, year)

    user_ids_rdd = reviews.map(lambda kv: kv['user_id'])
    result_dict['n_user'] = get_distinct_user(user_ids_rdd)

    business_user_rdd = reviews.map(lambda kv: (kv['business_id'], kv['user_id']))
    result_dict['top10_user'] = get_top_u_users(business_user_rdd, top_u)

    businesses_ids_rdd  = reviews.map(lambda kv: kv['business_id'])
    result_dict['n_business'] =get_distinct_businesses(businesses_ids_rdd)

    user_business_rdd = reviews.map(lambda kv: (kv['user_id'], kv['business_id']))
    result_dict['top10_business'] = get_top_b_businesses(user_business_rdd, top_b)

    #print(result_dict)

    with open(output_filepath, 'w+') as output_file:
        json.dump(result_dict, output_file, indent=2, separators=(',',':'))
    output_file.close()
